enginetest/evaluation.go

```diff
@@ -592,6 +598,37 @@ type CustomValueValidator interface {
 	Validate(interface{}) (bool, error)
 }
 
+// rowToSQL converts some expected row values into appropriate types according to given valid schema.
+// |convertTime| is true if it is a `SHOW` statements, except for `SHOW EVENTS`. This is set earlier in `checkResult()` method.
+func rowToSQL(s sql.Schema, expectedRow sql.Row, actualRow sql.Row, convertTime bool) (sql.Row, error) {
+	if s == nil || len(s) == 0 || types.IsOkResult(expectedRow) {
+		return expectedRow, nil
+	}
+	newRow := sql.Row{}
+	for i, v := range expectedRow {
+		_, isTime := v.(time.Time)
+		_, isStr := v.(string)
+		if v == nil {
+			newRow = append(newRow, nil)
+		} else if types.IsDecimal(s[i].Type) || (isTime && !convertTime) || (isStr && types.IsTextOnly(s[i].Type)) {
+			newRow = append(newRow, v)
+		} else if types.IsEnum(s[i].Type) || types.IsSet(s[i].Type) {
+			// engine tests use the index of the element, whereas over the wire tests use the string value
+			// TODO: need better check since the type received from go-sql-driver does not
+			//  provide data about its elements, we assume the result matches FOR NOW.
+			newRow = append(newRow, actualRow[i])
```

Comment:
Oh I added this when I was able to get the enum/set type information from go-sql-driver. Currently, this case cannot be detected because of the column type in the schema we receive as STRING type. I'll remove this as it's confusing.

Reply:
Ah... I see... I totally missed that the `Schema` we have is created from the result of go-sql-driver, not from the engine code. This limitation makes sense now.